ksqldb-serde/src/main/java/io/confluent/ksql/serde/connect/ConnectFormat.java

```diff
@@ -65,52 +74,100 @@ public boolean supportsSchemaInference() {
 
   @Override
   public List<SimpleColumn> toColumns(final ParsedSchema schema) {
-    final Schema connectSchema = toKsqlTransformer.apply(toConnectSchema(schema));
+    Schema connectSchema = toConnectSchema(schema);
 
-    return connectSchema.fields().stream()
+    if (connectSchema.type() != Type.STRUCT) {
+      if (!supportsFeature(SerdeFeature.UNWRAP_SINGLES)) {
+        throw new KsqlException("Schema returned from schema registry is anonymous type, "
+            + "but format " + name() + " does not support anonymous types. "
+            + "schema: " + schema);
+      }
+
+      connectSchema = SerdeUtils.wrapSingle(connectSchema);
+    }
+
+    final Schema rowSchema = toKsqlTransformer.apply(connectSchema);
+
+    return rowSchema.fields().stream()
         .map(ConnectFormat::toColumn)
         .collect(Collectors.toList());
   }
 
   public ParsedSchema toParsedSchema(
       final List<? extends SimpleColumn> columns,
-      final SerdeOptions serdeOptions,
+      final EnabledSerdeFeatures serdeFeatures,
       final FormatInfo formatInfo
   ) {
+    SerdeUtils.throwOnUnsupportedFeatures(serdeFeatures, supportedFeatures());
+
     final SchemaBuilder schemaBuilder = SchemaBuilder.struct();
     columns.forEach(col -> schemaBuilder.field(
         col.name().text(),
         SchemaConverters.sqlToConnectConverter().toConnectSchema(col.type()))
     );
 
-    final PersistenceSchema persistenceSchema =
-        buildValuePhysical(schemaBuilder.build(), serdeOptions);
+    final Schema schema = SerdeUtils.applySinglesUnwrapping(schemaBuilder.build(), serdeFeatures);
 
-    return fromConnectSchema(persistenceSchema.serializedSchema(), formatInfo);
+    return fromConnectSchema(schema, formatInfo);
   }
 
-  protected abstract Schema toConnectSchema(ParsedSchema schema);
+  @Override
+  public Serde<Struct> getSerde(
+      final PersistenceSchema schema,
+      final Map<String, String> formatProps,
+      final KsqlConfig config,
+      final Supplier<SchemaRegistryClient> srFactory
+  ) {
+    SerdeUtils.throwOnUnsupportedFeatures(schema.features(), supportedFeatures());
 
-  protected abstract ParsedSchema fromConnectSchema(Schema schema, FormatInfo formatInfo);
+    final ConnectSchema outerSchema = schema.connectSchema();
+    final ConnectSchema innerSchema = SerdeUtils
+        .applySinglesUnwrapping(outerSchema, schema.features());
 
-  private static PersistenceSchema buildValuePhysical(
-      final Schema valueConnectSchema,
-      final SerdeOptions serdeOptions
-  ) {
-    final boolean singleField = valueConnectSchema.fields().size() == 1;
+    final Class<?> targetType = SchemaConverters.connectToJavaTypeConverter()
+        .toJavaType(innerSchema);
 
-    final boolean unwrapSingle = serdeOptions.valueWrapping()
-        .map(option -> option == SerdeOption.UNWRAP_SINGLE_VALUES)
-        .orElse(false);
+    if (schema.features().enabled(SerdeFeature.UNWRAP_SINGLES)) {
```

Comment:
it seems a little weird to encapsulate this above (`applySinglesUnwrapping`) but then check for it explicitly here

Reply:
Why?  `applySinglesUnwrapping(outerSchema)` only extracts the inner schema. That's only half the storey. The code still needs to build a serde that can handle the unwrapping, i.e. extracting the value of the single column from the `Struct` passed to `serialize` and the reverse for `deserialize`.

Reply:
I feel like if I'm already checking the `UNWRAP_SINGLES` variable here, i might as well just extract the inner schema there as well. Not a biggie, but now we're checking it in two places and doing a no-op in one of them if it's disabled.